runtime/executor/src/main/java/org/apache/nemo/runtime/executor/task/MultiThreadParentTaskDataFetcher.java

```diff
@@ -127,14 +129,19 @@ private void fetchDataLazily() {
                   watermarkWithIndex.getIndex(), watermarkWithIndex.getWatermark());
               }
             } else {
+              if (element instanceof Finishmark) {
+                isFinishMarkProduced = true;
+              }
               // data element
               elementQueue.offer(element);
             }
           }
-
           // This iterator is finished.
           countBytesSynchronized(iterator);
-          elementQueue.offer(Finishmark.getInstance());
+          // If the current iterator hasn't produced finish mark, put it into the global queue.
+          if (!isFinishMarkProduced) {
```

Comment:
When are multiple finshmarks received?

Reply:
When the local output context closes, it produces a finish mark. In this case, the data fetcher receives a finish mark from the iterator. Then, the data fetcher shouldn't produce another finish mark when the iterator is finished.

Reply:
Then, receiving multiple Finishmark from an iterator is an exceptional case, so we should thrown an exception here?

Reply:
We shouldn't throw an exception. In the case of using ByteTransfer Context, its iterator doesn't produce a finish mark as an element at the end. But in the case of LocalTransfer context, its iterator produces a finish mark as an element at the end. So isFinishmarkProduced is to prevent producing two finish marks when the LocalTransfer context's iterator is finished.

Reply:
I fixed the LocalInputContext iterator so that it doesn't produce a finish mark at the end. We no longer need to handle the case in which multiple finish marks are produced.